### PR TITLE
local angular distributions

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

Add local angular distributions code in support of neutrino oscillations. This machinery is intended to find electron lepton number crossings. See, e.g., 
https://arxiv.org/abs/2404.17938

## PR Checklist

- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by Yurlungur** [comment]
Four species transport now supported. To use do the following:
1. Clone [my fork of NuLib](https://github.com/Yurlungur/NuLib/pull/1)
2. Modify the file `make_example_table.F90` and set `mytable_neutrino_scheme` to `2` and `number_output_species` to 4
3. build with `make -j`
4. run the resulting executable `make_example_table` 
5. `NuLib/scripts/table2bhlight.py` on the generated hdf5 file
6. Copy the generated table, which should end with `_bhlight.h5` to the location you wish to use for nubhlight
7. Modify the nublight `build.py` file or `param_template.dat` file to point to the file
8. In the nubhlight `build.py` file, either add this line `bhl.config.set_cparm("RAD_NUM_TYPES", 4)` or change `RAD_NUM_TYPES` to be 4.
9. build nubhlight and run

Below is the fornax 1 zone equilibrium test run with nue, nuebar, nux, nuxbar:
fornax-equilibrium-comparison

**Comment by Yurlungur** [comment]
important paper also for this MR https://arxiv.org/abs/2304.05044

File: core/decs.h
Comment (Yurlungur) [inline core/decs.h:1107]: needed for generalizing to 4 species transport

File: core/io.c
Comment (Yurlungur) [inline core/io.c:650]: these are diagnostics.

File: core/io.c
Comment (Yurlungur) [inline core/io.c:705]: diagnostics

File: core/io.c
Comment (Yurlungur) [inline core/io.c:1086]: diagnostic

File: core/io.c
Comment (Yurlungur) [inline core/io.c:1109]: diagnostics

File: core/io.c
Comment (Yurlungur) [inline core/io.c:1130]: A and B from Zaizan and Nakagura

File: core/io.c
Comment (Yurlungur) [inline core/io.c:1432]: auto formatting... oops but also 🤷

File: core/oscillations.c
Comment (Yurlungur) [inline core/oscillations.c:1]: meat of changes are here

File: core/oscillations.c
Comment (Yurlungur) [inline core/oscillations.c:48]: Used as a switch so we don't detect spurious crossings. Currently we're not outputting this to disk, but we could add it to the list of diagnostics.

File: core/oscillations.c
Comment (Yurlungur) [inline core/oscillations.c:73]: Payel's algorithm

File: core/oscillations.c
Comment (Yurlungur) [inline core/oscillations.c:121]: Crossing detector as defined by Zaizen and Nakamura

File: core/oscillations.c
Comment (Yurlungur) [inline core/oscillations.c:142]: A and B as defined by Zaizen and Nakamura. Currently using 0th moment. Should be straightforward to generalize.

File: core/oscillations.c
Comment (Yurlungur) [inline core/oscillations.c:163]: The magic.

File: prob/fornax_1zone/build.py
Comment (Yurlungur) [inline prob/fornax_1zone/build.py:1]: The changes here are maybe best left as intermediate tests. Should probably undo these changes before merge.

File: prob/multiscatt/build.py
Comment (Yurlungur) [inline prob/multiscatt/build.py:1]: The changes here are maybe best left as intermediate tests. Should probably undo these changes before merge.

File: prob/oscillations/build.py
Comment (Yurlungur) [inline prob/oscillations/build.py:1]: this test isn't ready yet.

File: prob/oscillations/problem.c
Comment (Yurlungur) [inline prob/oscillations/problem.c:1]: this test isn't ready yet.

File: prob/torus_cbc/build.py
Comment (Yurlungur) [inline prob/torus_cbc/build.py:1]: we should think about the changes here. We may wish to make a new build file rather than continuing to use the old one.

File: script/machine/darwin.py
Comment (Yurlungur) [inline script/machine/darwin.py:1]: this is my configuration. May or may not be wise to keep.

**Comment by Yurlungur** [comment]
This MR is ready. @kelslund @payelmuk150 please review.

**Comment by Yurlungur** [comment]
One-zone oscillation test
oscillations_1zone

File: script/analysis/hdf5_to_dict.py
Comment (kelslund) [inline script/analysis/hdf5_to_dict.py:118]: typo: read_hdr-var

Comment (Yurlungur) [reply script/analysis/hdf5_to_dict.py:118]: oops! Thank you!

File: core/decs.h
Comment (kelslund) [inline core/decs.h:1]: @Line 175: should this be implemented? i.e. 5 not 4?

File: README.md
Comment (kelslund) [inline README.md:1]: Just a heads up that line 1 of the parameters file (in NuLib build) should be modified to point to user-specified directory

File: core/decs.h
Comment (kelslund) [inline core/decs.h:385]: What does the [3] do here (line 385)?

File: core/diag.c
Comment (kelslund) [inline core/diag.c:591]: In decs.h, the condition is `#if RAD_NUM_TYPES >= 4`. Do you want to keep that for consistency here?

File: core/diag.c
Comment (kelslund) [inline core/diag.c:302]: What is the difference between these two conditions?

File: core/decs.h
Comment (Yurlungur) [reply core/decs.h:385]: I should potentially hardcode that to a named variable `LOCAL_NUM_MOMENTS` but it's basically `A`, `B` or `A - B`

File: core/diag.c
Comment (Yurlungur) [reply core/diag.c:591]: good idea

File: core/diag.c
Comment (Yurlungur) [reply core/diag.c:302]: One can compute G, A, B, etc without actually running the oscillations code. So `NEUTRINO_OSCILLATIONS` requires `LOCAL_ANGULAR_DISTRIBUTIONS` but not vice versa. Would be sufficient to include only the latter, but I was being careful.

**Comment by Yurlungur** [comment]
@kelslund thanks for the review! Have you made it through the whole thing? And if so, do you approve? (If so click the "finalize review" button on the top right of the diffs tab and click approve)

File: core/oscillations.c
Comment (kelslund) [inline core/oscillations.c:1]: I think this looks good!

Comment (Yurlungur) [reply core/oscillations.c:1]: :+1: Please click approve then :)

File: prob/torus_cbc/build.py
Comment (kelslund) [inline prob/torus_cbc/build.py:349]: Did we want to keep this after all?

File: prob/torus_cbc/build.py
Comment (kelslund) [reply prob/torus_cbc/build.py:428]: Did you also want that last line? Looks like it might be missing

File: prob/torus_cbc/build.py
Comment (Yurlungur) [reply prob/torus_cbc/build.py:349]: No I just forgot to revert it. Will do now.

Comment (Yurlungur) [reply prob/torus_cbc/build.py:349]: Just make sure to change it in the nulib fork too

File: prob/torus_cbc/build.py
Comment (Yurlungur) [reply prob/torus_cbc/build.py:428]: I don't know if I want to do that by default... let me add a flag

File: prob/torus_cbc/build.py
Comment (kelslund) [inline prob/torus_cbc/build.py:120]: What about subjecting this to an 'if oscillations'-type condition? If oscillations, NuLib, else, SFHo?

**Review by kelslund — APPROVED** [review]

File: prob/torus_cbc/build.py
Comment (Yurlungur) [reply prob/torus_cbc/build.py:120]: good idea will change that

File: core/decs.h
Comment (Yurlungur) [reply core/decs.h]: Renamed

**Pull request merged**
